[1단계 - 자동차 경주 구현] 짱구(신승철) 미션 제출합니다.

sc0116 opened the pull request:
안녕하세요 루피! 
저는 짱구에요!! 바쁘신데 귀한 시간 내주셔서 정말 감사합니다!!

우테코에서 첫 미션을 받았을 때, 프리코스 때 해봤던 미션이지만 생각처럼 되지 않아 아쉽네요ㅠㅠ  
개별 코드리뷰를 꼭 받고 싶었는데 제 코드가 너무 부끄럽네요...

페어와 미션을 진행하면서 궁금했던 점들을 정리해봤어요!

- 상수나 예외 메시지들을 그것들이 필요한 클래스 내부에 private static final 로 선언해두는 게 좋은지, 아니면 해당 종류들만 모아두는 클래스(예: ErrorMessage.class)를 만드는 게 좋은지 궁금합니다!!

- TDD를 처음 접하다보니 처음에 어느 정도까지 테스트 코드를 작성해야 하는지 잘 모르겠습니다... 물론 처음이다보니 어색해서 더 그럴 수 있다고 생각하지만 뭔가 TDD를 너무 많이 만드려고 페어와 고민하다보니 구현으로 넘어가기까지 오래걸리네요..!

- 혹시 테스트에서 private 메서드를 어떻게 테스트하는 것이 맞는지 모르겠습니다! 이번 미션에서는 private 메서드를 public으로 변경 후에 테스트를 돌렸습니다만, 이 방식이 맞다면, 테스트 성공 후에 다시 private으로 바꾸고 테스트 코드를 지워야 할까요?? 테스트가 중요하긴 하지만 테스트하려고 public으로 바꿔야 하나라는 생각이 들어서 여쭤봅니다!!

- 코드를 어느 정도까지 합쳐야? 할 지 모르겠습니다!
```java
private void result() {
    Winner winner = new Winner();
    List<String> winners = winner.getWinners(carController.getCars());

    outputView.printWinners(winners);
}
```

위 코드를 보면 메서드에서 우승자 리스트를 뜻하는 winners는 outputView에서 출력하기 위해 사용됩니다. 단 한번만 쓰이는데 저렇게 선언을 하고, 받아온 뒤, 출력 파라미터에 넘겨주는 게 의미가 있나 생각이 들어 아래와 같이 줄였습니다.

```java
private void result() {
    outputView.printWinners(new Winner().getWinners(carController.getCars());
}
```

하지만 이렇게 줄인 것이 과연 깔끔한 코드가 맞는지 의문이 들어 여쭤봅니다!!

---

아직 많이 부족하고 긴장이 되어 질문을 제대로 한건지 모르겠네요..!  
다시 한번 시간 내주셔서 감사합니다!!

sc0116 committed:
test(study): JUnit 학습 - String 테스트

sc0116 committed:
chore: ParameterizedTest 의존성 추가

sc0116 committed:
test(study): JUnit 학습 - Set 테스트

sc0116 committed:
test(calculator): 문자열 덧셈 계산기 테스트

sc0116 committed:
feat(calculator): 문자열 덧셈 계산기 구현

sc0116 committed:
docs: 기능목록 추가

sc0116 committed:
feat(racingcar): 자동차 이름 입력 UI 구현

sc0116 committed:
test(racingcar): 자동차 이름 입력 테스트

sc0116 committed:
feat(racingcar): 자동차 이름 입력 구현

sc0116 committed:
test(racingcar): 자동차 이름 입력 검증 테스트

sc0116 committed:
feat(racingcar): 자동차 이름 입력 검증 구현

sc0116 committed:
docs: 기능 목록 수정

자동차 입력 예외 변경

sc0116 committed:
feat(racingcar): 시도 횟수 입력 UI 구현

sc0116 committed:
test(racingcar): 자동차 이름 입력 테스트 수정

sc0116 committed:
test(racingcar): 시도 횟수 입력 테스트

sc0116 committed:
fix(racingcar): 자동차 이름 및 시도 횟수 인스턴스 변수로 변경

sc0116 committed:
feat(racingcar): 시도 횟수 입력 구현

sc0116 committed:
test(racingcar): 시도 횟수 입력 검증 테스트

sc0116 committed:
feat(racingcar): 시도 횟수 입력 검증 구현

sc0116 committed:
feat(racingcar): 실행 결과 UI 구현

sc0116 committed:
test(racingcar): 자동차 전진, 멈춤 테스트

sc0116 committed:
feat(racingcar): 시도 횟수만큼 자동차 전진 구현

sc0116 committed:
test(racingcar): 우승자 선정 테스트

sc0116 committed:
feat(racingcar): 우승자 선정 UI 구현

sc0116 committed:
feat(racingcar): 우승자 선정 구현

sc0116 committed:
feat(racingcar): 우승자 선정 구현

sc0116 committed:
refactor(racingcar): 자동차 경주 전체 리펙터링

- 가변 -> 불변 수정
- 변수명 수정
- indent 1 초과하는 부분 수정
- 테스트 코드 메서드 한글명 경고 제거

sc0116 committed:
refactor(racingcar): 자동차 경주 전체 리펙터링

- 가변 -> 불변 수정
- 변수명 수정
- indent 1 초과하는 부분 수정
- 테스트 코드 메서드 한글명 경고 제거

sc0116 committed:
refactor(racingcar): 변수 및 메서드 위치 일부 변경

workromancer commented on README.md:
```diff
@@ -1,7 +1,35 @@
+> 자동차 경주 미션 저장소
 
-자동차 경주 미션 저장소
+## 기능 목록
```
깔끔한 README 좋아요 👍

workromancer commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,40 @@
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(text);
+        if (m.find()) {
+            String customDelimiter = m.group(1);
+            return m.group(2).split(customDelimiter);
```
1과 2가 어떤 것을 의미하는지 명확하게 상수로 선언해서 표현하는 것은 어떨까요 ?

workromancer commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,40 @@
+
+    private static int sum(String[] tokens) {
+        int result = 0;
+        int number;
```
number 를 선언하면서 초기화가 되지 않고 있는데요.

for 문 내에서만 사용되는 것으로 보이는데 선언과 동시에 초기화하면 어떨까요 ?

workromancer commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,53 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class StringCalculatorTest {
```
해당 클래스의 테스트들이 스코프 내에서 checked exception 을 반환하지 않고 있는데요.

`throws Exception` 는 제거해도 괜찮지 않을까요 ?

workromancer commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,32 @@
+    private static final int MOVE_FORWARD_CONDITION = 4;
+
+    private final String name;
+    private int position = 0;
```
position 도 생성자에서 초기하는 것이 어떨까요 ?

workromancer commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,32 @@
+
+    public Car(final String name) {
+        this.name = name;
+    }
```
Car 의 생성자가 여러개 정의 되어있는데요. 하나의 생성자를 주 생성자로 정의하고 주 생성자를 반복하여 사용하는 것은 어떨까요 ?

https://jackjeong.tistory.com/148 이 글을 참고해보시면 좋을 것 같아요 ㅎㅎ

workromancer commented on src/main/java/racingcar/model/Winner.java:
```diff
@@ -0,0 +1,22 @@
+
+public class Winner {
+
+    public List<String> getWinners(final List<Car> cars) {
```
`List<Car> cars` 를 파라미터로 전달받지 않고 다른 방법으로 승자를 구할 수 있는 방법은 없을까요 ?

이렇게 파라미터로 전달받게 되면 `CarController` 에 getter 를 정의해서 getter 로 값을 꺼내 전달해야하지 않을까요 ?

view 로직 외에 getter 를 최소화할 수 있는 방법을 고민해보는 것은 어떨까요 ㅎㅎ

workromancer commented on src/main/java/racingcar/controller/CarController.java:
```diff
@@ -0,0 +1,36 @@
+
+import racingcar.model.Car;
+
+public class CarController {
```
해당 클래스의 역할이 모호한 것 같아요.

Controller 로 분류하는 것이 적절한지 한번 고민해봐주세요 ㅎㅎ

workromancer commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -0,0 +1,105 @@
+import racingcar.view.InputView;
+import racingcar.view.OutputView;
+
+public class GameController {
```
InputView 에서 할 수 있는 역할들이 해당 Controller 에 많이 섞여 있는 것 같아요 !

InputView 와 GameController 역할과 책임에 대하여 다시 한번 고민해보시는 것이 좋을 것 같아요

workromancer commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -0,0 +1,105 @@
+        return Arrays.stream(carNames)
+            .map(String::trim)
+            .toArray(String[]::new);
+    }
```
이런 로직들은 Controller 보다 View 로직에 가까운 로직이 아닐까요 ?

workromancer commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -0,0 +1,105 @@
+            .toArray(String[]::new);
+    }
+
+    private boolean CarNamesIsValidated(final String[] carNames) {
```
메서드 네임 컨벤션을 확인해주세요 !

workromancer commented on src/main/java/racingcar/controller/CarController.java:
```diff
@@ -0,0 +1,36 @@
+        return cars;
+    }
+
+    private int pickRandomNumber() {
```
랜덤 넘버가 아니라 자동차를 움직이는 숫자를 사용자 입력으로 받도록 변경해야 한다면 변경하기 어려워지지 않을까요 ?

workromancer commented on src/main/java/racingcar/utils/validator/CarNamesValidator.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.HashSet;
+import java.util.Set;
+
+public class CarNamesValidator {
```
Car 객체를 생성하기 전에 반드시 호출해줘야 하는 validate 메서드들로 보이는데요.

해당 validator 로직들을 Car 와 분리하신 이유가 궁금해요.

이렇게 분리하게 되면, Car 의 생성자로 생성할 때, name 을 5자 이상으로 전달하여 얼마든지 객체를 생성할 수 있게 될텐데요.

Car 의 name 이 5자 미만일때 Car 를 반드시 생성하지 않도록 해야한다면, 생성자에서 확인하는 것이 더 좋지 않을까요 ?

sc0116 replied:
헉 Gmail로 리뷰를 봐서 댓글을 달 수 있는지 너무 늦게 알았네요ㅠㅠㅠㅠ  

일단 CarNamesValidator 클래스는 경주 시작 전 경주에 참가할 자동차 목록을 받았을 때, 이 이름들이 올바른 이름인지 검증하기 위해 만들었습니다..!

해당 로직을 Car와 분리한 이유는 입력한 이름들이 올바른지 검증한 후, 만약 올바르다면 자동차 이름 목록을 넘겨줘서 Car 객체로 생성하려 했어서 분리했습니다.

하지만 아직 잘 이해가 가지 않습니다..! 이름 목록을 입력받고 검증하는 것이 아니라 목록을 넘겨서 생성해줄 때 검증을 하게 되면 너무 늦게 검증한다는 생각이 듭니다..! 혹시 이름 목록을 입력받을 때, 생성할 때 두 번 검증해야 하는건가요??

workromancer replied:
오.. 저도 늦게 봤네요..

Input 을 검증하는 것과 객체 생성을 검증하는 것이 동일한 것 일까요 ?
한번 고민해봐주시면 좋을 것 같아요 ㅎㅎ

workromancer commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.Scanner;
+
+public class InputView {
```
GameController 에서 수행하고 있는 로직들과 InputView 의 로직이 적절하게 재배치되면 좋을 것 같아요 !

workromancer reviewed (changes requested):
안녕하세요 짱구
리뷰어 루피입니다. 🤠

긴장하셨다고 하셔서 소개를 덧붙여보자면, 저도 3년 전에 우아한테크코스를 수강한 수강생인데요 ㅎㅎ 최대한 긴장하지 마시고 저에게 많은 것들을 얻어가주세요 !

첫번째 미션을 하면서 많은 것들이 새로우셨을텐데 고생 많으셨습니다.

작성해주신 코드에 대하여 몇 가지 코멘트를 남겨두었는데요.

다음 리뷰 요청 주실 때, 충분히 고민해보시고 반영 부탁드려요 ㅎㅎ

진행 중에 고민되는 포인트들이 있다면 얼마든지 DM 이나 코멘트 남겨주세요.

아래는 리뷰 요청 주실 때 남겨주신 질문들에 대한 답변입니다.

1. 상수나 예외 메시지들을 그것들이 필요한 클래스 내부에 private static final 로 선언해두는 게 좋은지, 아니면 해당 종류들만 모아두는 클래스(예: ErrorMessage.class)를 만드는 게 좋은지 궁금합니다!!

상수로 선언하는 것이 어떤 의미를 가지는지 고민해보시면 좋을 것 같아요.
어떤 값을 의미있는 이름으로 표현하기 위해 상수로 선언하는 경우가 많은데요. 이 의미있는 이름의 유효 범위에 따라 private 으로 할 지 public 으로 할 지 선택하는 편입니다.
단순히 이런 값들만 모아두는 클래스를 정의하는 것은 큰 의미가 없을 것 같네요 ㅎㅎ


2. TDD를 처음 접하다보니 처음에 어느 정도까지 테스트 코드를 작성해야 하는지 잘 모르겠습니다... 물론 처음이다보니 어색해서 더 그럴 수 있다고 생각하지만 뭔가 TDD를 너무 많이 만드려고 페어와 고민하다보니 구현으로 넘어가기까지 오래걸리네요..!

뭐든 처음하면 어색하기 마련이죠 ㅎㅎ 앞으로 레벨 1을 하시는 동안 차차 적응해나가실 수 있을꺼예요.
처음에 감을 잡으시면서는 요구사항에 있는 내용들 위주로 TDD 를 연습해보시는 것이 좋을 것 같아요. 그리고 TDD 의 핵심 중에 하나는 "매우 짧은 개발 사이클을 반복하는 것" 이라고 생각하는데요.
구현으로 넘어가기까지 너무 오래 걸리신다면 좀 더 작은 단위로 TDD 를 하시는 범위를 줄여보시는 것이 좋을 것 같습니다.
TFD (Test First Development) + Refactoring 을 반복하는 것이 TDD 이라고 강의시간에 들으셨을 것 같은데요.
요구사항에 맞는 테스트를 잘게 쪼개서 테스트를 작성하시고, 바로 이 테스트 코드를 통과시키기 위한 프로덕션 코드를 구현해보세요 ㅎㅎ


3. 혹시 테스트에서 private 메서드를 어떻게 테스트하는 것이 맞는지 모르겠습니다! 이번 미션에서는 private 메서드를 public으로 변경 후에 테스트를 돌렸습니다만, 이 방식이 맞다면, 테스트 성공 후에 다시 private으로 바꾸고 테스트 코드를 지워야 할까요?? 테스트가 중요하긴 하지만 테스트하려고 public으로 바꿔야 하나라는 생각이 들어서 여쭤봅니다!!

우리가 테스트 코드를 작성하는 이유는 프로덕션 코드를 작성하는데 도움을 얻기 위해서 인데요. 반대로 테스트를 위해 프로덕션 코드를 변경하거나 작성하는 것은 지양하는 것이 좋습니다.
public 메서드를 통해 실행되지 않는 private 메서드는 불필요한 메서드일텐데요. public 메서드를 테스트 하신다면 private 메서드도 수행될테니 별도로 private 을 public 으로 열어서 테스트를 작성하실 필요는 없습니다.

개인적으로 private 메서드가 지나치게 많아지는 경우는 해당 클래스의 역할과 책임에 대하여 한번 더 고민해보는데요. private 메서드가 지나치게 많아지는 것을 클래스가 지나치게 많은 역할과 책임을 하고 있다는 증거가 될 수도 있다고 생각하고 있습니다. 


4. 코드를 어느 정도까지 합쳐야? 할 지 모르겠습니다!

질문의 글자수는 짧은데, 답변을 드리기에는 하루 이틀도 부족할 것 같습니다... ㅎㅎ
이번 미션에서 감을 잡을 수 있도록 최대한 도와드리도록 할께요.
질문 주신 코드에 대한 코멘트도 남겨두었으니 확인해보세요

sc0116 committed:
refactor(calculator): 문자열 덧셈 계산기 리팩터링

- 매직 넘버 -> 상수 변경
- 변수 사용 범위에 맞게 선언 위치 변경
- 하나의 테스트 메서드에서 하나의 테스트만 수행하도록 변경
- 메소드명 일부 수정

sc0116 committed:
refactor(racingcar): 자동차 경주 리펙터링

- 난수 생성 기능 클래스로 분리
- 우승자 관련 로직 RacingGame 클래스로 이동

sc0116 committed:
refactor(racingcar): 자동차 경주 리펙터링

- 입력 검증 로직 View 부분으로 위치 이동
- Car 클래스 필드 생성자에서 초기화

sc0116 committed:
refactor(racingcar): tryCount getter 대신 시도 횟수를 판별하는 메서드 추가

workromancer commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,49 @@
+
+public class StringCalculator {
+
+    public static final int CUSTOM_DELIMITER_INDEX = 1;
+    public static final int TEXT_INDEX = 2;
```
의도가 명확하게 드러나서 읽기 좋네요 👍

다만, 클래스 밖에서 상수가 사용되지 않고 있는 것 같아요~
접근 제어자를 한번 확인해주세요~

workromancer commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -0,0 +1,56 @@
+    private final List<Car> cars;
+    private int tryCount;
+
+    public RacingGame(String[] cars, int tryCount) {
+        this.cars = new ArrayList<>();
+        createCars(cars);
+        this.tryCount = tryCount;
+    }
+
+    private void createCars(final String[] carNames) {
+        Arrays.stream(carNames)
+            .forEach(carName -> cars.add(new Car(carName)));
+    }
```
cars 를 초기화 한 뒤에 조작하고 있는데요.
생성과 초기화를 동시에 할 수는 없을까요 ?

sc0116 replied:
제 코드는 루피 말씀대로 new ArrayList<>()로 생성을 한 후 createCars 메서드로 초기화를 하네요...

지금 떠오른 아이디어는 createCars 메서드의 반환형이 void인데 List<Car>로 바꾸고 스트림의 map과 collect를 이용해서 cars로 넘겨주면 될 것 같아요!!

더 생각하고 리뷰 요청했어야 하는데 죄송합니당...

workromancer replied:
아녀유... 왜 죄송해하십니까... 우테코는 토론, 설득하며 좋은 방향을 찾아나가는 교육이지... 다르게 생각하면 혼나고 점수깍고 그런곳이 아닙니다 !!

workromancer commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -0,0 +1,56 @@
+
+import racingcar.utils.RandomNumber;
+
+public class RacingGame {
```
적절하게 잘 분리해주신 것 같아요 👍

workromancer commented on src/main/java/racingcar/utils/RandomNumber.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.util.Random;
+
+public class RandomNumber {
```
난수 생성의 범위가 1-10 이 아닌 5-25 로 변경된다면 변경이 어려울 수 있지 않을까요 ?

workromancer commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,49 @@
+        int result = 0;
+
+        for (String token : tokens) {
+            int number = Integer.parseInt(token);
```
👍

workromancer reviewed (approved):
짱구, 안녕하세요.
이전 리뷰를 잘 반영해 주신 부분도 있고, 확인이 어려운 부분도 있어서 확인된 부분은 Resolve 처리를 확인이 어려운 부분은 그대로 두었어요~
해결된 내용은 Resolve 처리 부탁드려요 
추가로 코멘트를 남겨두었으니 그 부분도 확인 부탁드려요 !
추가 코멘트와 아직 Resolve 처리가 되지 않은 부분들은 2단계에서 확인 부탁드려요~
첫 미션 고생 많으셨습니다 !
